src/core/qgsarrowiterator.cpp

```diff
@@ -0,0 +1,403 @@
+/***************************************************************************
+    qgsarrowreader.h
+    ---------------------
+    begin                : November 2025
+    copyright            : (C) 2025 by Dewey Dunnington
+    email                : dewey at dunnington dot ca
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsarrowiterator.h"
+
+#include <nanoarrow/nanoarrow.hpp>
+
+#include "qgsfeatureiterator.h"
+#include "qgsvectorlayer.h"
+
+#define QGIS_NANOARROW_THROW_NOT_OK_ERR( expr, err )                                                                      \
+  do                                                                                                                      \
+  {                                                                                                                       \
+    const int ec = ( expr );                                                                                              \
+    if ( ec != NANOARROW_OK )                                                                                             \
+    {                                                                                                                     \
+      throw QgsException( QString( "nanoarrow error (%1): %2" ).arg( ec ).arg( QString::fromUtf8( ( err )->message ) ) ); \
```

Comment:
```suggestion
      throw QgsException( QStringLiteral( "nanoarrow error (%1): %2" ).arg( ec ).arg( QString::fromUtf8( ( err )->message ) ) ); \
```

(don't ask, it's a silly Qt thing!)